### PR TITLE
Flipbook: Add play/pause, speed control, and image toolbar features

### PR DESCRIPTION
**Pull request by goplayoutside3**
Based on https://github.com/zooniverse/front-end-monorepo/pull/3782 's branch.
~~This is a draft PR because additional unit tests are needed~~

## Package
lib-classifier

## Linked Issue and/or Talk Post
Flipbook Project Board: https://github.com/zooniverse/front-end-monorepo/projects/21

## Describe your changes
- Added the relevant functions passed from subject viewer store to enable pan, zoom, rotation, invert colors.
- Added a play/pause button that loops through the subject's images.
- While focused on the subject image, the spacebar key will play/pause too.
- Added a speed control dropdown with 5 options based on [Invision Design](https://projects.invisionapp.com/prototype/Flipbook-Viewer-cl7ar8qa402m83i01w6uvj5hq)
- (Note that this PR does not fully handle play iterations lab config yet)

## How to Review
To see the Flipbook Viewer run locally, visit the following projects.

[Wildwatch Burrowing Owl](https://local.zooniverse.org:3000/projects/sandiegozooglobal/wildwatch-burrowing-owl/classify/workflow/21141?env=production)
[A Flipbook Test Project](https://local.zooniverse.org:3000/projects/goplayoutside3/fem-flipbook-viewer/classify/workflow/22532?env=production)

You can also see the Flipbook component in the classifier's storybook.

# Checklist

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

### CODE REVIEW

**Comment by eatyourgreens** [comment]
After #3782 merged, the history of this branch had diverged from `master`. I'm not quite sure how that happened, but I've cleaned up the history of this branch and removed all the commits that were already included in #3782 and #3777.

**Comment by eatyourgreens** [comment]
You could use the `prefers-reduced-motion` media query to check whether or not to autoplay the slideshow.
```js
const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');

if (!mediaQuery || mediaQuery.matches) {
  doSomethingWithoutAnimation();
} else {
  doSomethingWithAnimation();
}
```

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/FlipbookViewerContainer.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/FlipbookViewerContainer.js:31]: Good catch - I'll remove it

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:133]: `width` works as a Grommet attribute, but `textAlign` only works as a style object attribute

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: I'll move these to a `styled()` class, except for the attributes dependent on `activeFrame` and `thumbnailUrl`

File: packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js:9]: `playIterations` comes from panoptes as a string. An empty string like `''` means infinite iterations, otherwise the value is something like `"3"`

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: You can pass props to `styled()`. We’ve got a few examples of dynamic classes in our code but the documentation is here:
https://styled-components.com/docs/basics#adapting-based-on-props

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: I think Grommet uses `align` for `text-align` but it’s been a while since I’ve looked that up.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: Here’s an example from our code where the colour of an icon depends on a variable. 
https://github.com/zooniverse/front-end-monorepo/blob/de2ca76167873d3de90aa9970dc05fab032b5cf3/packages/lib-react-components/src/FavouritesButton/HeartIcon.js#L5-L7

Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: More info about the `css` helper here: https://github.com/zooniverse/front-end-monorepo/pull/1496

The classifier doesn’t use SSR so you can probably ignore it, except for nested template literals.

File: packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js:9]: I wonder why this is camel case when snake case is used for other attributes. That’s confusing. Probably a project builder bug, that we can’t change now that the value is set on live workflows.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:18]: HTML buttons can’t contain `div` tags so this component is generating invalid HTML somewhere. React will warn about invalid DOM nesting errors when this happens.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:17]:
```suggestion
const DirectionButton = styled(Button)`
  border: ‘none’;
  padding: 0;
```

You can add your border and padding here, or use Grommet’s `border` and `pad` props on the buttons.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:30]:
```suggestion
  background-position: center;

  &[aria-selected=true] {
    border: solid 2px ${props => props.theme.global.colors['neutral-2']};
  }
```

You can use an attribute selector to add a border to the active button.

**Comment by eatyourgreens** [comment]
The viewer flashes when the images first load, because each image is loaded when the active frame changes. PFE avoids this by preloading all three frames in the background before playing.

This is a bug in the `useSubjectImage` hook at the moment, which can only preload one image URL. It needs to be updated to preload an array of URLs.


https://user-images.githubusercontent.com/59547/199956908-6c4da8ac-865c-444f-afba-bfee071ceb92.mov

**Comment by eatyourgreens** [comment]
After setting the player speed, I have to select it again for each new subject. I think the player should remember the selected speed, otherwise we'll get complaints from volunteers who prefer slower or faster playback speeds.

**Comment by eatyourgreens** [comment]
When I select the subject, then press the spacebar, the slideshow starts playing but the page scrolls down. You'll need to call `event.preventDefault()` to stop that from happening. Maybe also cancel event bubbling with `event.stopPropagation()`.


https://user-images.githubusercontent.com/59547/199958590-c964fb67-db81-49fa-b1eb-76bf52a5fc81.mov

**Comment by eatyourgreens** [comment]
If I tab into the thumbnail buttons, while the player is playing, then press the arrow keys, the player tries to both respond to the arrow keys and keep playing. It should maybe stop playing when the arrow keys are used to navigate the frames.

There's also a bug where focus doesn't change when I use the arrow keys. The active frame changes, but the focused button doesn't change. The [MultiFrame viewer](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/story/subject-viewers-multiframeviewer--with-less&globals=locale:en;locales.en:English;locales.test:Test+Language) doesn't have this bug. That bug could be caught by adding some tests for `document.activeElement` when the selected frame is being selected from the keyboard.

**Review by eatyourgreens — CHANGES_REQUESTED** [review]
This is looking good, but the UX still feels like it needs some improvements before it's ready.

File: packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js
Comment (eatyourgreens) [reply packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js:9]: You could store the value as a number here, which might be more intuitive. Maybe see the default value to `Infinity` to make the code easier to follow.

Comment (goplayoutside3) [reply packages/lib-classifier/src/store/WorkflowStore/Workflow/WorkflowConfiguration/WorkflowConfiguration.js:9]: Agreed. This PR doesn't fully handle play iterations, it just assumes infinite so I'll note this ideas a possible change for my next PR.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: You're right, the `css` helper can be used here and I've updated accordingly.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:18]: The invalid HTML is a result of Grommet's Button handling an `icon` and `label` prop. Would you recommend creating my own button component instead?

**Comment by goplayoutside3** [comment]
> After setting the player speed, I have to select it again for each new subject. I think the player should remember the selected speed, otherwise we'll get complaints from volunteers who prefer slower or faster playback speeds.

Good point! PFE does this (i.e [Wildwatch Burrowing Owl](https://www.zooniverse.org/projects/sandiegozooglobal/wildwatch-burrowing-owl/classify)). I added this as a to-do card in [Flipbook Project Board](https://github.com/zooniverse/front-end-monorepo/projects/21) so that I can dedicate a small scoped PR to the change.

**Comment by goplayoutside3** [comment]
> EDIT: a quick preloader might look something like this:
> 
> ```js
> function preloadSubject({ locations }) {
>   locations.forEach(location => {
>     const [url] = Object.values(location)
>     const img = new Image()
>     img.src = url
>   }
> }
> ```
 
I've also added this as an item card to the [Flipbook Project Board](https://github.com/zooniverse/front-end-monorepo/projects/21). Do you recommend `preloadSubject()` happens in `useSubjectImage()` or in FlipbookViewer?

**Comment by goplayoutside3** [comment]
> When I select the subject, then press the spacebar, the slideshow starts playing but the page scrolls down. 

> If I tab into the thumbnail buttons, while the player is playing, then press the arrow keys, the player tries to both respond to the arrow keys and keep playing.

Thanks for catching - this behavior should be fixed!

**Comment by goplayoutside3** [comment]
> There's also a bug where focus doesn't change when I use the arrow keys. The selected frame changes, but the focused button doesn't change. 

I modified `handleKeyDown()` to change focus to the selected thumbnail button.

**Comment by goplayoutside3** [comment]
I think this is ready for review again!

**Review by eatyourgreens — APPROVED** [review]

**Comment by eatyourgreens** [comment]
Changes look good! I've got minor comments but nothing blocking this PR.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:109]:
```suggestion
  const handleThumbnailClick = (index) => {
```

Typo here.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:193]: Not needed for this PR, but the thumbnail service will fail for URLs with origins that aren't owned by Zooniverse. We should add a check that the origin is `https://panoptes-uploads.zooniverse.org`, and fall back to using the full size image for projects that host images on their own domains.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:211]:
```suggestion
                      onClick={() => handleThumbnailClick(index)}
```
Typo.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js]: It's not a big deal. HTML parsers can handle invalid button content, probably by ignoring the invalid block layout. Different browsers might handle a `<div>` inside a button differently, so the layout isn't guaranteed to be the same cross-browser.

The bug is caused by this Grommet `Box`. One way to avoid it is to explicitly pass children to the button, rather than `icon` and `label`. We can also open an issue with Grommet, and a PR to fix this.
https://github.com/grommet/grommet/blob/f9c351e1bb7d55a18f94cc60c017fac50d26f7e3/src/js/components/Button/Button.js#L339-L357

**Comment by eatyourgreens** [comment]
>I've also added this as an item card to the [Flipbook Project Board](https://github.com/zooniverse/front-end-monorepo/projects/21). Do you recommend preloadSubject() happens in useSubjectImage() or in FlipbookViewer?

I think preloading in the Flipbook Viewer would work for now, so that there aren't side effects for other viewers and projects.

One other way of implementing a preloader would be to add a second `useSubjectImage` hook, so that you load the next image, for `frame + 1`, while displaying the active frame.

**Comment by eatyourgreens** [comment]
Not related to the changes here, but a couple of things I noticed while testing:
- when the subject image has focus, the keyboard controls for pan and zoom aren't working.
- when I'm using the speed dropdown from the keyboard, the low colour contrast makes it impossible to tell which choice is active (and will fail an automated colour contrast check.)

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js
Comment (goplayoutside3) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/FlipbookViewer/components/FlipbookControls.js:193]: Good to note! I added this to the Flipbook Project Board

**Comment by goplayoutside3** [comment]
> when I'm using the speed dropdown from the keyboard, the low colour contrast makes it impossible to tell which choice is active (and will fail an automated colour contrast check.)

That's a good point - I changed the background color to `accent-1`

> when the subject image has focus, the keyboard controls for pan and zoom aren't working.

This was an easy fix! I didn't realize `withKeyZoom()` was used for this.

**Pull request merged**
